vendor/github.com/pstibrany/services/basic_service.go

```diff
@@ -0,0 +1,341 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// StartingFn is called when service enters Starting state. If StartingFn returns
+// error, service goes into Failed state. If StartingFn returns without error, service transitions into
+// Running state (unless context has been canceled).
+//
+// serviceContext is a context that is finished at latest when service enters Stopping state, but can also be finished
+// earlier when StopAsync is called on the service. This context is derived from context passed to StartAsync method.
+type StartingFn func(serviceContext context.Context) error
+
+// RunningFn function is called when service enters Running state. When it returns, service will move to Stopping state.
+// If RunningFn or Stopping return error, Service will end in Failed state, otherwise if both functions return without
+// error, service will end in Terminated state.
+type RunningFn func(serviceContext context.Context) error
+
+// StoppingFn function is called when service enters Stopping state. When it returns, service moves to Terminated or Failed state,
+// depending on whether there was any error returned from previous RunningFn (if it was called) and this StoppingFn function. If both return error,
+// RunningFn's error will be saved as failure case for Failed state.
+type StoppingFn func() error
```

Comment:
In general, there is one `Service` interface, and currently one `BasicService` implementation.

These functions are specific to the implementation. `BasicService` requires all of them, but not all services need to use them all. These functions are not related to different types of services, single service often uses all three.

Reply:
Yes, right. I was suggesting to define an interface for each service type with the methods that the caller needs to define to create a service like for `InitIdleService` make an interface as below:
```
type IdleService interface {
	type StartingFn func(serviceContext context.Context) error
	type StoppingFn func() error
}
```

Reply:
But even "idle" service may not require starting or stopping function, and using interface would force it to have one (which does nothing).

Furthermore, by using interface, one would need to have type implementing the interface, but that's often not practical. For example, `initQuerier` now returns idle service without needing any extra type: https://github.com/pstibrany/cortex/blob/5c9d7faad7121d982ed16cd11a601d693e258b99/pkg/cortex/modules.go#L323

And if for example Ingester implemented this interface directly, those `StartingFn` and `StoppingFn` would become part of its public API, which is what I want to avoid.